platform-sdk/docs/proposals/TSS-Block-Signing/tss-block-signing.md

```diff
@@ -0,0 +1,286 @@
+
+    ...
+
+    /**
+     * Requests a ledger signature on a message hash.  The ledger signature is computed asynchronously and returned
+     * to all consumers that have been registered through {@link #registerLedgerSignatureConsumer(Consumer)}.
+     *
+     * @param messageHash The hash of the message to be signed by the ledger.
+     */
+    void requestLedgerSignature(@Nonnull byte[] messageHash);
+
+    /**
+     * Registers a consumer of pairs where the first element is the message hash and the second element is the
+     * ledger signature on the message hash.
+     *
+     * @param consumer the consumer of ledger signatures on message hashes.
+     */
+    void registerLedgerSignatureConsumer(@Nonnull final Consumer<Pair<Bytes, PairingSignature>> consumer);
+}
+```
+
+#### System Transactions
+
+The `TssShareSignatureTransaction` is a new system transaction that is only generated by the `TssBaseService` and
+not user generated. It may be possible that a node is aggregating share signature transactions from
+multiple ledger signature requests. In addition to the `share_signature`, the `roster_hash` and `share_index` are
+needed to identify the share public key to use for validation. The `message_hash` is needed to identify the message
+that was signed. A ledger signature can be produced by a threshold number of `share_signatures` for the same
+message. The threshold value for aggregation is determined by the roster indicated by the `roster_hash`.
+
+```protobuf
+message TssShareSignatureTransaction {
+  /**
+   * The hash of the roster containing the node whose share produced the share signature.
+   */
+  bytes roster_hash = 1;
+
+  /**
+   * The index of the share that produced this share signature.
+   */
+  uint64 share_index = 2;
+
+  /**
+   * The hash of the message that was signed by the share.
+   */
+  bytes message_hash = 3;
+
+  /**
+   * The share signature produced by the share.
+   */
+  bytes share_signature = 4;
+}
+```
+
+### Architecture and/or Components
+
+In addition to the extension of the `TssBaseService` interface, the service's `TransactionHandler` is extended to
+handle `TssShareSignatureTransaction` in the `pre-handle` phase.
+
+### Core Behaviors
+
+Thread Execution:
+
+* All code related to requesting a ledger signature,
+  generating share signature transactions, and handling share signature transactions do not modify the state and
+  should not execute on the `transaction handling thread`.
+* As soon as the ledger signature for a message can be constructed, all registered consumers are notified with the
+  message hash and the ledger signature. If there are multiple consumers registered, whether they are given the
+  signature serially or in parallel makes no difference as the data is immutable and consumption is expected to be
+  idempotent.
+
+Ledger Signature Generation:
+
+* The storage of `TssShareSignatureTransaction` is only in memory and expires after a configured amount of time.
```

Comment:
If I understand the description of the proposed implementation correctly, then we expect the code on every node to call the request method at approximately the same time, and that would cause each node to generate a `TssShareSignatureTransaction` and send it to every other node, thus allowing each node to collect the full set of the signatures.

In other words, there's not a way for a single node to call the request method and get a random message signed with the ledger id because there's not a way for that node to request all the other nodes to produce their signatures. So this means that we wouldn't be able to expose the signing method as an API on a node.

Is it good or is it bad?

Reply:
Right, this is an internal API that has to be called deterministically on all nodes.  If we wanted to open up the service, like a Notary that signs user/customer messages, then we would need to create a new NotaryService that is user facing and create a new user generated transaction request for that.  It would get gossiped to all nodes and the NotaryService would handle the new transaction after consensus in a deterministic call to create the ledger signature by registering a ledger signature consumer with the TssBaseService and invoking the `requestLedgerSignature` method.

Reply:
Makes sense. Thanks for the additional details!